### PR TITLE
Create a ReversibleTransactionExecutor that can take any sender without their private key

### PR DESCRIPTION
**Pull request by tinchou**

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 160 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 118 major
* ![MINOR][MINOR] 15 minor
* ![INFO][INFO] 26 info

#### Top 10 issues

1. ![BLOCKER][BLOCKER] [BlockChainImpl.java#L192](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L192): Catch Exception instead of Throwable. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1181)
1. ![MAJOR][MAJOR] [RskFactory.java#L83](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L83): Method has 14 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L117](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L117): Constructor new co.rsk.core.bc.BlockChainImpl(Repository, BlockStore, ReceiptStore, PendingState, EthereumListener, AdminInfo, BlockValidator) makes call to non-final method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APCOA_PARTIALLY_CONSTRUCTED_OBJECT_ACCESS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L199](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L199): The Cyclomatic Complexity of this method "internalTryToConnect" is 22 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L305](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L305): Method co.rsk.core.bc.BlockChainImpl.internalTryToConnect(Block) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L359](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L359): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L370](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L370): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L375](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L375): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L445](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L445): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L523](https://github.com/rsksmart/rskj/blob/56d5a2f0cb01dfa394ccc66d8f017118c018f49f/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L523): Method co.rsk.core.bc.BlockChainImpl.flushData() orders expressions in a conditional in a sub optimal way [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ASEO_SUBOPTIMAL_EXPRESSION_ORDER)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 160 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 118 major
* ![MINOR][MINOR] 15 minor
* ![INFO][INFO] 26 info

#### Top 10 issues

1. ![BLOCKER][BLOCKER] [BlockChainImpl.java#L192](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L192): Catch Exception instead of Throwable. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1181)
1. ![MAJOR][MAJOR] [RskFactory.java#L83](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L83): Method has 14 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L117](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L117): Constructor new co.rsk.core.bc.BlockChainImpl(Repository, BlockStore, ReceiptStore, PendingState, EthereumListener, AdminInfo, BlockValidator) makes call to non-final method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APCOA_PARTIALLY_CONSTRUCTED_OBJECT_ACCESS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L199](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L199): The Cyclomatic Complexity of this method "internalTryToConnect" is 22 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L305](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L305): Method co.rsk.core.bc.BlockChainImpl.internalTryToConnect(Block) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L359](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L359): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L370](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L370): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L375](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L375): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L445](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L445): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L523](https://github.com/rsksmart/rskj/blob/a3252584bb56b14ddef901be52de2fcce1d8d5c4/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L523): Method co.rsk.core.bc.BlockChainImpl.flushData() orders expressions in a conditional in a sub optimal way [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ASEO_SUBOPTIMAL_EXPRESSION_ORDER)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/org/ethereum/core/ReversibleTransactionExecutor.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.ethereum.core;
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/core/ReversibleTransactionExecutor.java]: Move to co.rsk

**Comment by tinchou** [comment]
Approved but blocked by https://github.com/rsksmart/rskj/pull/223

**Pull request merged**